### PR TITLE
feat: discord name synchronisation

### LINKED ISSUE #5391: Dicord Link Nicknames
### Feature description

Through the Discord Link module, allow for Minecraft nicknames to be set to sync with the corresponding Discord nickname.

For example, if a user's Discord server nickname is "James", then when they link their Minecraft and Discord account via the Discord Link module, their Minecraft nickname will update to "James" as well.

### How the feature is useful

This would save Minecraft server admins a lot of time, rather than manually updating each player's nickname (when we can't allow users to do it themselves).

### PR DESCRIPTION
**Pull request by gbaranski**
### Information

This PR closes #5391. 

### Details

**Proposed feature:**  

Automatic synchronisation of Discord and Minecraft names.
Can work in two directions:
- MC -> Discord
- Discord -> MC

**Environments tested:**    

OS: Linux

Java version: 18

- [x] Most recent Paper version (1.XX.Y, git-Paper-BUILD)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

### CODE REVIEW

**Comment by gbaranski** [comment]
As for now it doesn't really work, I'm not sure how to set nicknames 

https://github.com/gbaranski/Essentials/blob/dfdc0cf112c66a846e9a30b73325fe438f917cfb/EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java#L41

```java
ess.getServer().dispatchCommand(Bukkit.getConsoleSender(), String.format("nickname %s %s", user.getName(), member.getEffectiveName()));
```

Does anyone have any idea how else it could be solved? Maybe somehow internally by sending NickChangeEvent?

File: EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java [hunk 1]
```diff
@@ -27,6 +32,20 @@ public LinkBukkitListener(EssentialsDiscordLink ess) {
         this.ess = ess;
     }
 
+    private void updateName(IUser user, InteractionMember member) {
+        switch (ess.getSettings().getNameSyncDirection()) {
+            case DISCORD_TO_MC:
+                // TODO: Remove this log
+                ess.getLogger().info(String.format("Setting username of %s to %s", user.getName(), member.getEffectiveName()));
+                // TODO: use a better way to change the display name
+                ess.getServer().dispatchCommand(Bukkit.getConsoleSender(), String.format("nickname %s %s", user.getName(), member.getEffectiveName()));
```
Comment (Octol1ttle) [inline EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java:41]: Use the same code used in `/nick`: https://github.com/EssentialsX/Essentials/blob/2.x/Essentials/src/main/java/com/earth2me/essentials/commands/Commandnick.java#L110-L117

File: EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java [hunk 2]
```diff
@@ -52,7 +71,12 @@ public void onConnect(final AsyncPlayerPreLoginEvent event) {
             return;
```
Comment (Octol1ttle) [inline EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java:71]: The nickname synchronization code will not run due to this `return` if the link policy isn't set to kick unlinked players

File: EssentialsDiscordLink/src/main/resources/config.yml [hunk 3]
```diff
@@ -50,3 +50,9 @@ role-sync:
   roles:
     22222222222222222: vip
     33333333333333333: booster
+
+# In what mode names should be synchronised. Available modess:
```
Comment (Octol1ttle) [inline EssentialsDiscordLink/src/main/resources/config.yml]:
```suggestion
# In what mode names should be synchronised. Available modes:
```

File: EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/DiscordLinkSettings.java [hunk 4]
```diff
@@ -68,6 +69,8 @@ private Map<String, String> _getRoleSyncRoles() {
         return config.getStringMap("role-sync.roles");
     }
 
+    public NameSyncDirection getNameSyncDirection() { return nameSyncDirection; }
```
Comment (Octol1ttle) [inline EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/DiscordLinkSettings.java]:
```suggestion
    public NameSyncDirection getNameSyncDirection() {
        return nameSyncDirection;
    }
```

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: EssentialsDiscordLink/src/main/resources/config.yml [hunk 5]
```diff
@@ -50,3 +50,9 @@ role-sync:
   roles:
     22222222222222222: vip
     33333333333333333: booster
+
+# In what mode names should be synchronised. Available modes:
+# none - Name synchronisation disabled
+# mc-to-discord - Minecraft names will be set on Discord
+# discord-to-mc - Discord names will be set on Minecraft
+name-sync: none
```
Comment (untuned) [inline EssentialsDiscordLink/src/main/resources/config.yml:58]: I would probably mention the fact that the server owner will not be able to be synchronized, as bots cannot modify their nickname.

**Pull request closed without merging**